### PR TITLE
Bitmap encoding

### PR DESCRIPTION
**Pull request by raver119**
***WIP; DO NOT MERGE;***

This PR adds few tweaks for Spark distributed training: 
- bitmap encoding for Spark
- sparse -> dense and dense -> sparse encodings auto switch
- shake option, to send updates with smaller thresholds
- threshold decay option

### CODE REVIEW

File: deeplearning4j-nn/src/main/java/org/deeplearning4j/optimize/solvers/accumulation/CudaGradientsAccumulator.java [hunk 1]
```diff
@@ -224,7 +224,12 @@ public void applyUpdate(StepFunction function, INDArray params, INDArray updates
             while (!messages.get(index.get()).isEmpty()) {
                 INDArray compressed = messages.get(index.get()).poll();
 
-                INDArray decoded = Nd4j.getExecutioner().thresholdDecode(compressed, updates);
+                int encoding = compressed.data().getInt(3);
+                if (encoding == 0)
```
Comment (AlexDBlack) [inline deeplearning4j-nn/src/main/java/org/deeplearning4j/optimize/solvers/accumulation/CudaGradientsAccumulator.java]: Minor style point - replace 0 magic constant with a named constant - ```THRESHOLD_ENCODE_IDX``` or something.

File: deeplearning4j-nn/src/main/java/org/deeplearning4j/optimize/solvers/accumulation/EncodingHandler.java [hunk 2]
```diff
@@ -39,7 +49,18 @@ public EncodingHandler(double threshold) {
...
 
     public EncodingHandler(double threshold, Double boundary) {
+        this(threshold, threshold, 0.0, 0, boundary);
+    }
+
+    public EncodingHandler(double threshold, double minThreshold, double thresholdStep,  int shakeFrequency) {
+        this(threshold, minThreshold, thresholdStep, shakeFrequency, null);
+    }
+
+    public EncodingHandler(double threshold, double minThreshold, double thresholdStep, int shakeFrequency, Double boundary) {
```
Comment (AlexDBlack) [inline deeplearning4j-nn/src/main/java/org/deeplearning4j/optimize/solvers/accumulation/EncodingHandler.java]: Javadoc for args needed here

File: deeplearning4j-nn/src/main/java/org/deeplearning4j/optimize/solvers/accumulation/EncodingHandler.java [hunk 3]
```diff
@@ -56,14 +77,57 @@ public void initialize(@NonNull GradientsAccumulator accumulator) {
...
+                double encodingRatio = encLen * 100.0 / updates.length();
+
+                // if updates are too dense - we fallback to bitmap encoding
+                if (encLen > (updates.lengthLong() / 16 + 5))
+                    bitmapMode.get().set(true);
+
+
+                // after encoding is finished, and updates are sparse enough - let's step down a bit
+                // and we don't step down too early, so we wait for 50 iterations at least to step down
+                if (minThreshold <= currentThreshold.get().get() && minThreshold < currentThreshold.get().get() - thresholdStep && iterations.get().get() > lastStep.get().get() + 50 && encodingRatio < 0.05) {
```
Comment (AlexDBlack) [inline deeplearning4j-nn/src/main/java/org/deeplearning4j/optimize/solvers/accumulation/EncodingHandler.java]: Hardcoded threshold/constants here - seems to be reasonable, but should this perhaps be configurable?

File: deeplearning4j-scaleout/spark/dl4j-spark-parameterserver/src/main/java/org/deeplearning4j/spark/parameterserver/networking/SilentTrainingDriver.java [hunk 4]
```diff
@@ -139,7 +142,19 @@ public void startTraining(SilentUpdatesMessage message) {
             // master invokes everything, since that's Silent Worker approach: we want master to be always up-to-date
             synchronized (this) {
                 // threshold decoder is inplace & fast
-                Nd4j.getExecutioner().thresholdDecode(message.getUpdates(), updates);
+                int encoding = message.getUpdates().data().getInt(3);
+                if (encoding == 0) {
+                    Nd4j.getExecutioner().thresholdDecode(message.getUpdates(), updates);
+                    sparseCounter.incrementAndGet();
+                } else if (encoding == 1) {
```
Comment (AlexDBlack) [inline deeplearning4j-scaleout/spark/dl4j-spark-parameterserver/src/main/java/org/deeplearning4j/spark/parameterserver/networking/SilentTrainingDriver.java]: Magic constants

File: deeplearning4j-scaleout/spark/dl4j-spark-parameterserver/src/main/java/org/deeplearning4j/spark/parameterserver/networking/SilentTrainingDriver.java [hunk 5]
```diff
@@ -139,7 +142,19 @@ public void startTraining(SilentUpdatesMessage message) {
...
+                    Nd4j.getExecutioner().bitmapDecode(message.getUpdates(), updates);
+                    denseCounter.incrementAndGet();
+                }
+
+                if ((sparseCounter.get() + denseCounter.get()) % 100 == 0) {
+                    log.info("Sparse/Dense ratio: {}", String.format("%.2f", (sparseCounter.get() +1) / (double) (denseCounter.get() + 1)));
```
Comment (AlexDBlack) [inline deeplearning4j-scaleout/spark/dl4j-spark-parameterserver/src/main/java/org/deeplearning4j/spark/parameterserver/networking/SilentTrainingDriver.java]: Debug level for the log?
In the future (UI pass) we might look at tracking this info over time...

File: deeplearning4j-scaleout/spark/dl4j-spark-parameterserver/src/main/java/org/deeplearning4j/spark/parameterserver/networking/WiredEncodingHandler.java [hunk 6]
```diff
@@ -32,6 +32,14 @@ public WiredEncodingHandler(double threshold, Double boundary) {
         super(threshold, boundary);
     }
 
+    public WiredEncodingHandler(double threshold, double minThreshold, double thresholdStep,  int shakeFrequency) {
+        this(threshold, minThreshold, thresholdStep, shakeFrequency, null);
+    }
+
+    public WiredEncodingHandler(double threshold, double minThreshold, double thresholdStep, int shakeFrequency, Double boundary) {
```
Comment (AlexDBlack) [inline deeplearning4j-scaleout/spark/dl4j-spark-parameterserver/src/main/java/org/deeplearning4j/spark/parameterserver/networking/WiredEncodingHandler.java]: Javadoc

File: deeplearning4j-scaleout/spark/dl4j-spark-parameterserver/src/main/java/org/deeplearning4j/spark/parameterserver/training/SharedTrainingMaster.java [hunk 7]
```diff
@@ -1077,6 +1087,51 @@ public Builder updatesThreshold(double threshold) {
         }
 
         /**
+         * Once update with given threshold become too sparse, threshold will be decreased by thresholdStep, but now below minimum threshold
```
Comment (AlexDBlack) [inline deeplearning4j-scaleout/spark/dl4j-spark-parameterserver/src/main/java/org/deeplearning4j/spark/parameterserver/training/SharedTrainingMaster.java]: Typo? ```but now``` -> ```but not```

File: deeplearning4j-scaleout/spark/dl4j-spark-parameterserver/src/main/java/org/deeplearning4j/spark/parameterserver/training/SharedTrainingMaster.java [hunk 8]
```diff
@@ -1077,6 +1087,51 @@ public Builder updatesThreshold(double threshold) {
...
+         * @param threshold
+         * @return
+         */
+        public Builder minUpdatesThreshold(double threshold) {
+            this.minThreshold = threshold;
+            return this;
+        }
+
+        /**
+         * Step size for threshold decay
```
Comment (AlexDBlack) [inline deeplearning4j-scaleout/spark/dl4j-spark-parameterserver/src/main/java/org/deeplearning4j/spark/parameterserver/training/SharedTrainingMaster.java]: Note default value here so users have an idea of the scale we're talking about?

File: deeplearning4j-nn/src/main/java/org/deeplearning4j/optimize/solvers/accumulation/EncodingHandler.java (hunk 2 above)
Comment (raver119) [inline deeplearning4j-nn/src/main/java/org/deeplearning4j/optimize/solvers/accumulation/EncodingHandler.java]: Done

File: deeplearning4j-nn/src/main/java/org/deeplearning4j/optimize/solvers/accumulation/EncodingHandler.java (hunk 3 above)
Comment (raver119) [inline deeplearning4j-nn/src/main/java/org/deeplearning4j/optimize/solvers/accumulation/EncodingHandler.java]: Done

File: deeplearning4j-scaleout/spark/dl4j-spark-parameterserver/src/main/java/org/deeplearning4j/spark/parameterserver/networking/SilentTrainingDriver.java (hunk 5 above)
Comment (raver119) [inline deeplearning4j-scaleout/spark/dl4j-spark-parameterserver/src/main/java/org/deeplearning4j/spark/parameterserver/networking/SilentTrainingDriver.java]: Removed. That's for my debugging...

File: deeplearning4j-scaleout/spark/dl4j-spark-parameterserver/src/main/java/org/deeplearning4j/spark/parameterserver/networking/WiredEncodingHandler.java (hunk 6 above)
Comment (raver119) [inline deeplearning4j-scaleout/spark/dl4j-spark-parameterserver/src/main/java/org/deeplearning4j/spark/parameterserver/networking/WiredEncodingHandler.java]: Done

File: deeplearning4j-scaleout/spark/dl4j-spark-parameterserver/src/main/java/org/deeplearning4j/spark/parameterserver/training/SharedTrainingMaster.java (hunk 7 above)
Comment (raver119) [inline deeplearning4j-scaleout/spark/dl4j-spark-parameterserver/src/main/java/org/deeplearning4j/spark/parameterserver/training/SharedTrainingMaster.java]: fixed

File: deeplearning4j-scaleout/spark/dl4j-spark-parameterserver/src/main/java/org/deeplearning4j/spark/parameterserver/training/SharedTrainingMaster.java (hunk 8 above)
Comment (raver119) [inline deeplearning4j-scaleout/spark/dl4j-spark-parameterserver/src/main/java/org/deeplearning4j/spark/parameterserver/training/SharedTrainingMaster.java]: Fixed

File: deeplearning4j-nn/src/main/java/org/deeplearning4j/optimize/solvers/accumulation/CudaGradientsAccumulator.java (hunk 1 above)
Comment (raver119) [inline deeplearning4j-nn/src/main/java/org/deeplearning4j/optimize/solvers/accumulation/CudaGradientsAccumulator.java]: Fixed

**Pull request merged**